atc/api/configserver/save.go

```diff
@@ -165,48 +168,66 @@ func (s *Server) SaveConfig(w http.ResponseWriter, r *http.Request) {
 }
 
 // Simply validate that the credentials exist; don't do anything with the actual secrets
-func validateCredParams(vars creds.Variables, config atc.Config, session lager.Logger) error {
+func validateCredParams(credMgrVars creds.Variables, config atc.Config, session lager.Logger) error {
 	var errs error
 
 	for _, resourceType := range config.ResourceTypes {
-		_, err := creds.NewSource(vars, resourceType.Source).Evaluate()
+		_, err := creds.NewSource(credMgrVars, resourceType.Source).Evaluate()
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		}
 	}
 
 	for _, resource := range config.Resources {
-		_, err := creds.NewSource(vars, resource.Source).Evaluate()
+		_, err := creds.NewSource(credMgrVars, resource.Source).Evaluate()
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		}
 
-		_, err = creds.NewString(vars, resource.WebhookToken).Evaluate()
+		_, err = creds.NewString(credMgrVars, resource.WebhookToken).Evaluate()
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		}
 	}
 
 	for _, job := range config.Jobs {
 		for _, plan := range job.Plan {
-			_, err := creds.NewParams(vars, plan.Params).Evaluate()
+			_, err := creds.NewParams(credMgrVars, plan.Params).Evaluate()
 			if err != nil {
 				errs = multierror.Append(errs, err)
 			}
 
-			if plan.TaskConfig != nil {
-				if plan.TaskConfig.ImageResource != nil {
-					_, err = creds.NewSource(vars, plan.TaskConfig.ImageResource.Source).Evaluate()
-					if err != nil {
-						errs = multierror.Append(errs, err)
-					}
+			if plan.TaskConfigPath != "" {
+				// external task - we can't really validate much right now, because task yaml will be
+				// retrieved in runtime during job execution. but we can validate vars and params which will be
+				// passed to this task
+				err = creds.NewTaskParamsValidator(credMgrVars, plan.TaskConfig.Params).Validate()
+				if err != nil {
+					errs = multierror.Append(errs, err)
 				}
 
-				_, err = creds.NewTaskParams(vars, plan.TaskConfig.Params).Evaluate()
+				err = creds.NewTaskVarsValidator(credMgrVars, plan.TaskVars).Validate()
+				if err != nil {
+					errs = multierror.Append(errs, err)
+				}
+
+			} else if plan.TaskConfig != nil {
```

Comment:
:thinking: Supporting `vars:` with `config:` feels like more trouble than it's worth, because it's not really obvious which vars will be satisfied at `set-pipeline` time and which vars will be satisfied by `vars:` at runtime. In fact I could see this shooting devs in the foot if a var in `set-pipeline` unintentionally takes precedence over one that's also specified in `vars:`.

I think it would be simpler to only permit `vars:` with `file:`. This would also reduce the dependency between on `/exec` here which feels a little more coupled than usual:

```
~/w/concourse (master) $ ag /exec atc/api/
~/w/concourse (master) $
```

Reply:
agreed. I don't know how to fully remove dependency on exec(), but agree that we shouldn't support vars for embedded tasks. will fix